### PR TITLE
Render parameters in same group as object in the form

### PR DESCRIPTION
**Pull request by sverhoeven**
Refs #25

### CODE REVIEW

File: src/store.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { atom, selector, useRecoilState, useRecoilValue } from 'recoil'
+import { atom, selector, SetterOrUpdater, useRecoilState, useRecoilValue, useSetRecoilState } from 'recoil'
```
Comment (Peter9192) [inline src/store.ts:1]: The [explain like I'm 5](https://www.youtube.com/watch?v=U9XStcquQyY) video on the recoil home page couldn't have been more obscure...

File: src/catalog.ts [hunk 2]
```diff
@@ -8,10 +9,12 @@ export async function fetchCatalog (catalogUrl: string): Promise<ICatalog> {
     throw new Error('Error retrieving catalog')
   }
   const body = await response.text()
-  const catalog = load(body)
-  if (!isCatalog(catalog)) {
+  const unGroupedcatalog = load(body)
```
Comment (Peter9192) [inline src/catalog.ts:12]:
```suggestion
  const ungroupedCatalog = load(body)
```

File: README.md [hunk 3]
```diff
@@ -101,6 +101,23 @@ The catalog is a YAML formatted file which tells the app what nodes are availabl
...
+
+```json
+{
+    "prop1": {
+        "ui:group": "<group name>"
+    }
+}
+```
+
+Any property with a group uiSchema field will be grouped together with other properties with the same group in the form. The generate toml file will have the props ungrouped.
```
Comment (Peter9192) [inline README.md:119]:
```suggestion
Any property with a group uiSchema field will be grouped together with other properties with the same group in the form. The generated toml file will have the props ungrouped.
```

**Review by Peter9192 — APPROVED** [review]
Awesome! Perhaps change the PR name to be a bit less abstract? Something like "Allow grouping of related parameters"?

File: src/grouper.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+import { ICatalog, IParameters } from './types'
+import { JSONSchema7 } from 'json-schema'
+import { isObject } from './utils/isObject'
+
+// TODO call when loading catalog and store in catalog.node.grouped...
+export function groupSchema (schema: JSONSchema7, uiSchema: UiSchema): JSONSchema7 {
+  const newSchema = JSON.parse(JSON.stringify(schema))
+
+  Object.entries(uiSchema).forEach(([k, v]) => {
+    // TODO recursivly, now only loops over first direct props
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply src/grouper.ts:11]: I do not need it for haddock3., It is there as a reminder that when the builder is used to build for another program some more work is needed.

File: src/store.ts (hunk 1 above)
Comment (sverhoeven) [reply src/store.ts:1]: Sorry to hear that. When I was picking a state manager I watched https://www.youtube.com/watch?v=P95DuIBwnqw&t=50s

File: src/catalog.ts (hunk 2 above)
Comment (sverhoeven) [reply src/catalog.ts:12]: nice catch

**Comment by sverhoeven** [comment]
Thanks for reviewing.

File: src/store.ts (hunk 1 above)
Comment (Peter9192) [reply src/store.ts:1]: Luckily the [other video](https://www.youtube.com/watch?v=_ISAA_Jt9kI&feature=emb_logo) on their home page was much better:

**Pull request merged**